cgrok/tree.c

```diff
@@ -0,0 +1,198 @@
+  TCTREE *tree = malloc(sizeof(TCTREE));
+  tree->dict = hb_dict_new(cp, tcfree);
+  tree->iter = NULL;
+  return tree; 
+}
+
+// Iterate over elements in ascending key order
+void tctreeiterinit(TCTREE *tree) {
+  if (tree->iter != NULL) {
+    dict_itor_free(tree->iter);
+  }
+  tree->iter = dict_itor_new(tree->dict);
+  dict_itor_first(tree->iter);
+}
+
+const void *tctreeiternext(TCTREE *tree, int *sp) {
+  void *key = dict_itor_key(tree->iter);
+  dict_itor_next(tree->iter);
+  *sp = 0;
+  if (key) {
+    *sp = *(uint32_t*)key;
+    return key+4;
+  }
+  return NULL;
+}
+
+void *tcpack(const void *value, uint32_t size) {
+  char *buf = malloc(size + 5);
+  memcpy(buf, &size, 4);
+  memcpy(buf+4, value, size);
+  buf[size+4] = 0;
+  return buf;
+}
+
+void tctreeput(TCTREE *tree, const void *kbuf, int ksiz, const void *vbuf, int vsiz) {
+  void *key = tcpack(kbuf, ksiz);
+  void *val = tcpack(vbuf, vsiz);
+  bool inserted; 
+  void ** valPtr = dict_insert(tree->dict, key, &inserted);
+  *valPtr = val;
+}
+
+bool tctreeputkeep(TCTREE *tree, const void *kbuf, int ksiz, const void *vbuf, int vsiz) {
+  void *key = tcpack(kbuf, ksiz); 
+  void *val = tcpack(vbuf, vsiz); 
+  bool inserted;
+  void **valPtr = dict_insert(tree->dict, key, &inserted);
+  if (inserted) {
+    *valPtr = val;
+  }
+  return inserted;
+}
+
+const void *tctreeget(TCTREE *tree, const void *kbuf, int ksiz, int *sp) {
+  const void *key = tcpack(kbuf, ksiz);
+  void *value = dict_search(tree->dict, key);
+  *sp = 0;
+  if (value) {
+    *sp = *(uint32_t*)value;
+    return value+4;
+  }
+  return NULL;
+}
+
+void tctreeclear(TCTREE *tree) {
+  dict_clear(tree->dict); 
+}
+
+void tctreedel(TCTREE *tree) {
+  dict_free(tree->dict);
+  if (tree->iter) {
+    dict_itor_free(tree->iter);
+  }
+  free(tree);
+}
+
+// A simple, singly linked list
+TCLIST *tclistnew(void) {
+  TCLIST *list = malloc(sizeof(TCLIST));
+  list->len = 0;
+  TCLISTELEM *newElem = malloc(sizeof(TCLISTELEM));
+  newElem->elem = 0;
+  list->head = newElem;
+  return list;
+}
+
+int tclistnum(const TCLIST *list) {
+  return list->len;
+}
+
+void *tclistpack(void *buf, uint32_t size) {
+  void *elem = malloc(size+1);
+  memcpy(elem, buf, size);
+  ((char *)elem)[size] = 0;
+  return elem;
+}
+
+void tclistpush(TCLIST *list, const void *ptr, int size) {
+  TCLISTELEM *head = list->head;
+  for (int i=0; i < (list->len); i++) {
+    head = head->next;
+  }
+  TCLISTELEM *newElem = malloc(sizeof(TCLISTELEM));
+  newElem->elem = tclistpack(ptr, size);
+  newElem->size = size;
+  head->next = newElem;
+  list->len += 1;
+}
+
+void *tclistremove(TCLIST *list, int index, int *sp) {
+  if (index >= list->len) {
+    return NULL;
+  }
+  TCLISTELEM *head = list->head;
+  for (int i=0; i < index; i++) {
+    head = head->next;
+  }
+  
+  TCLISTELEM *removed = head->next;
+  head->next = removed->next;
+  void* elem = removed->elem;
+  *sp = removed->size;
+  free(removed);
+  list->len -= 1;
+  return elem;
+}
```

Comment:
The TC docs are a big ball of crazy. This is what it says about tclistremove: 

```
Because an additional zero code is appended at the end of the region of the return value, the return value can be treated as a character string. Because the region of the return value is allocated with the `malloc' call, it should be released with the `free' call when it is no longer in use. If `index' is equal to or more than the number of elements, no element is removed and the return value is `NULL'.
```

I take this to mean "When you give us a value, we malloc sizeof(value)+1 and copy your value there with a NULL terminator. When you ask for it back we give you our copy, and freeing it is your problem. " But I could be wrong